### ISSUE TITLE
[Question]Can embedding weights shared between encoder and decoder?

### ISSUE THREAD

**Issue by SefaZeng** [NVIDIA/FasterTransformer#483]
I wanna deploy a translation model with FasterTransformer. But I found in the example the encoder embedding, decoder embedding and decoder output projection are using separate parameters. But sometimes these 3 embeddings is shared during training, and it could save a lot of GPU memory.

Can Faster-Transformer share all the 3 embeddings?

**Comment by byshiue** [comment]
For pytorch op, the weights are pass by users, so you can pass same pointer for these three weights. 

For cpp or triton backend, it needs to modify the source codes and hard to share the weights between encoder and decoder because they are separate modules. Some decoder modules (like gpt, t5) have share the weights on embedding and decoder output projection if they are same.

**Comment by SefaZeng** [comment]
> For pytorch op, the weights are pass by users, so you can pass same pointer for these three weights.
> 
> For cpp or triton backend, it needs to modify the source codes and hard to share the weights between encoder and decoder because they are separate modules. Some decoder modules (like gpt, t5) have share the weights on embedding and decoder output projection if they are same.

Great! Could you tell me where the decoder parameter sharing is implemented? I'd like to try it on decoding.

**Comment by SefaZeng** [comment]
> For pytorch op, the weights are pass by users, so you can pass same pointer for these three weights.
> 
> For cpp or triton backend, it needs to modify the source codes and hard to share the weights between encoder and decoder because they are separate modules. Some decoder modules (like gpt, t5) have share the weights on embedding and decoder output projection if they are same.

As far as I can see, the cpp code for gpt and t5 just uses the same parameter file for input and output projection, but still consumes twice as much memory?

**Comment by byshiue** [comment]
> Great! Could you tell me where the decoder parameter sharing is implemented? I'd like to try it on decoding.

It is only implemented in decoding because decoder doses not contain the projection and embedding. 

> As far as I can see, the cpp code for gpt and t5 just uses the same parameter file for input and output projection, but still consumes twice as much memory?

I don't know how you get this conclusion.

**Comment by SefaZeng** [comment]
> > Great! Could you tell me where the decoder parameter sharing is implemented? I'd like to try it on decoding.
> 
> It is only implemented in decoding because decoder doses not contain the projection and embedding.
> 
> > As far as I can see, the cpp code for gpt and t5 just uses the same parameter file for input and output projection, but still consumes twice as much memory?
> 
> I don't know how you get this conclusion.

Thanks for your reply!

**Comment by SefaZeng** [comment]
> > Great! Could you tell me where the decoder parameter sharing is implemented? I'd like to try it on decoding.
> 
> It is only implemented in decoding because decoder doses not contain the projection and embedding.
> 
> > As far as I can see, the cpp code for gpt and t5 just uses the same parameter file for input and output projection, but still consumes twice as much memory?
> 
> I don't know how you get this conclusion.

Hi, @byshiue. Hope you are still here. I build an NMT module in `src/fastertransformer/th_op` and it includes both encoder and decoding. The compilation goes well. But when I try to load the libth_nmt.so to python, it says torch can't find `torch.classes.FasterTransformer.NMT` but can find `torch.classes.FasterTransformerNMT`.
Did I miss something? Should I build another NMT model in this directory `src/fastertransformer/model`?

**Comment by byshiue** [comment]
We suggestion writing something like
```python
        try:
            self.decoding = torch.classes.FasterTransformer.T5Decoding(head_num, head_size, inter_size, mem_d_model,
                                                                       d_model, num_layer,
                                                                       vocab_size, num_bucket, num_expert, max_distance,
                                                                       q_scaling, start_id, end_id,
                                                                       tensor_para_size, pipeline_para_size,
                                                                       t5_with_bias,
                                                                       position_embedding_type, moe_k, activation_type,
                                                                       tie_word_embeddings, adapter_inter_size,
                                                                       adapter_norm_position,
                                                                       moe_layer_index, *decoding_weight_list)
        except:
            self.decoding = torch.classes.FasterTransformerT5Decoding(head_num, head_size, inter_size, mem_d_model,
                                                                      d_model, num_layer,
                                                                      vocab_size, num_bucket, num_expert, max_distance,
                                                                      q_scaling, start_id, end_id,
                                                                      tensor_para_size, pipeline_para_size,
                                                                      t5_with_bias,
                                                                      position_embedding_type, moe_k, activation_type,
                                                                      tie_word_embeddings, adapter_inter_size,
                                                                      adapter_norm_position,
                                                                      moe_layer_index, *decoding_weight_list)
```
and register the op like https://github.com/NVIDIA/FasterTransformer/blob/main/src/fastertransformer/th_op/t5/T5DecodingOp.cc#L702-L706

**Comment by SefaZeng** [comment]
> We suggestion writing something like
> 
> ```python
>         try:
>             self.decoding = torch.classes.FasterTransformer.T5Decoding(head_num, head_size, inter_size, mem_d_model,
>                                                                        d_model, num_layer,
>                                                                        vocab_size, num_bucket, num_expert, max_distance,
>                                                                        q_scaling, start_id, end_id,
>                                                                        tensor_para_size, pipeline_para_size,
>                                                                        t5_with_bias,
>                                                                        position_embedding_type, moe_k, activation_type,
>                                                                        tie_word_embeddings, adapter_inter_size,
>                                                                        adapter_norm_position,
>                                                                        moe_layer_index, *decoding_weight_list)
>         except:
>             self.decoding = torch.classes.FasterTransformerT5Decoding(head_num, head_size, inter_size, mem_d_model,
>                                                                       d_model, num_layer,
>                                                                       vocab_size, num_bucket, num_expert, max_distance,
>                                                                       q_scaling, start_id, end_id,
>                                                                       tensor_para_size, pipeline_para_size,
>                                                                       t5_with_bias,
>                                                                       position_embedding_type, moe_k, activation_type,
>                                                                       tie_word_embeddings, adapter_inter_size,
>                                                                       adapter_norm_position,
>                                                                       moe_layer_index, *decoding_weight_list)
> ```
> 
> and register the op like https://github.com/NVIDIA/FasterTransformer/blob/main/src/fastertransformer/th_op/t5/T5DecodingOp.cc#L702-L706

Thanks for you reply! I have build an module called NMT in `src/fastertransformer/th_op/nmt` and it holds a `ftencoder` and a `ftdecoding`. I can get the right results when I give the fp32 parameter to it. But if I turn to parameter to fp16, the `ftencoder` return tensor with all NAN value.  Can you give some advice on what could have gone wrong?

**Comment by SefaZeng** [comment]
> 

Here is NMTOp.cc
```
#include "src/fastertransformer/th_op/nmt/NMTOp.h"

namespace th = torch;

namespace torch_ext {
using torch::Tensor;

FasterTransformerNMT::FasterTransformerNMT(int64_t head_num,
                                                     int64_t size_per_head,
                                                     int64_t inter_size,
                                                     int64_t mem_hidden_dim,
                                                     int64_t layer_num,
                                                     int64_t vocab_size,
                                                     int64_t start_id,
                                                     int64_t end_id,
                                                     double beam_search_diversity_rate,
                                                     int64_t top_k,
                                                     double top_p,
                                                     double temperature,
                                                     double len_penalty,
                                                     double repetition_penalty,
                                                     th::Tensor self_layernorm_gamma,
                                                     th::Tensor self_layernorm_beta,
                                                     th::Tensor self_kernel_q,
                                                     th::Tensor self_bias_q,
                                                     th::Tensor self_output_kernel,
                                                     th::Tensor self_output_bias,
                                                     th::Tensor cross_layernorm_gamma,
                                                     th::Tensor cross_layernorm_beta,
                                                     th::Tensor cross_kernel_q,
                                                     th::Tensor cross_kernel_k,
                                                     th::Tensor cross_kernel_v,
                                                     th::Tensor cross_bias_q,
                                                     th::Tensor cross_bias_k,
                                                     th::Tensor cross_bias_v,
                                                     th::Tensor cross_output_kernel,
                                                     th::Tensor cross_output_bias,
                                                     th::Tensor ffn_layernorm_gamma,
                                                     th::Tensor ffn_layernorm_beta,
                                                     th::Tensor inter_kernel,
                                                     th::Tensor inter_bias,
                                                     th::Tensor output_kernel,
                                                     th::Tensor output_bias,
                                                     th::Tensor decoding_gamma,
                                                     th::Tensor decoding_beta,
                                                     th::Tensor embedding_table,
                                                     th::Tensor position_encoding_table,
                                                     th::Tensor embedding_kernel,
                                                     th::Tensor embedding_bias,
                                                     th::Tensor pre_attr_layernorm_gamma,
                                                     th::Tensor pre_attr_layernorm_beta,
                                                     th::Tensor q_kernel,
                                                     th::Tensor q_bias,
                                                     th::Tensor k_kernel,
                                                     th::Tensor k_bias,
                                                     th::Tensor v_kernel,
                                                     th::Tensor v_bias,
                                                     th::Tensor attr_output_kernel,
                                                     th::Tensor attr_output_bias,
                                                     th::Tensor pre_ffn_layernorm_gamma,
                                                     th::Tensor pre_ffn_layernorm_beta,
                                                     th::Tensor enc_inter_kernel,
                                                     th::Tensor enc_inter_bias,
                                                     th::Tensor enc_output_kernel,
                                                     th::Tensor enc_output_bias,
                                                     th::Tensor post_transformer_layernorm_gamma,
                                                     th::Tensor post_transformer_layernorm_beta,
                             int64_t enc_layer_num,
                                                      bool remove_padding,
                                                      bool allow_gemm_test,
                                                      bool sparse,
                                                      double q_scaling):
    _st(self_layernorm_gamma.scalar_type()),
    _remove_padding(remove_padding),
    weights_enc{pre_attr_layernorm_gamma,
            pre_attr_layernorm_beta,
            q_kernel,
            q_bias,
            k_kernel,
            k_bias,
            v_kernel,
            v_bias,
            attr_output_kernel,
            attr_output_bias,
            pre_ffn_layernorm_gamma,
            pre_ffn_layernorm_beta,
            enc_inter_kernel,
            enc_inter_bias,
            enc_output_kernel,
            enc_output_bias,
            post_transformer_layernorm_gamma,
            post_transformer_layernorm_beta},
    weights_dec{self_layernorm_gamma,    self_layernorm_beta,   self_kernel_q,        self_bias_q,    self_output_kernel,
            self_output_bias,        cross_layernorm_gamma, cross_layernorm_beta, cross_kernel_q, cross_kernel_k,
            cross_kernel_v,          cross_bias_q,          cross_bias_k,         cross_bias_v,   cross_output_kernel,
            cross_output_bias,       ffn_layernorm_gamma,   ffn_layernorm_beta,   inter_kernel,   inter_bias,
            output_kernel,           output_bias,           decoding_gamma,       decoding_beta,  embedding_table,
            position_encoding_table, embedding_kernel,      embedding_bias}

{
    std::cout<<"NMT init"<<std::endl;
    CHECK_INPUT(self_layernorm_gamma, _st);     // layer_num, hidden_dim
    CHECK_INPUT(self_layernorm_beta, _st);      // layer_num, hidden_dim
    CHECK_INPUT(self_kernel_q, _st);            // layer_num, hidden_dim, 3 * hidden_dim
    CHECK_INPUT(self_bias_q, _st);              // layer_num, 3 * hidden_dim
    CHECK_INPUT(self_output_kernel, _st);       // layer_num, hidden_dim, hidden_dim
    CHECK_INPUT(self_output_bias, _st);         // layer_num, hidden_dim
    CHECK_INPUT(cross_layernorm_gamma, _st);    // layer_num, hidden_dim
    CHECK_INPUT(cross_layernorm_beta, _st);     // layer_num, hidden_dim
    CHECK_INPUT(cross_kernel_q, _st);           // layer_num, hidden_dim, hidden_dim
    CHECK_INPUT(cross_kernel_k, _st);           // layer_num, mem_hidden_dim, hidden_dim
    CHECK_INPUT(cross_kernel_v, _st);           // layer_num, mem_hidden_dim, hidden_dim
    CHECK_INPUT(cross_bias_q, _st);             // layer_num, hidden_dim
    CHECK_INPUT(cross_bias_k, _st);             // layer_num, hidden_dim
    CHECK_INPUT(cross_bias_v, _st);             // layer_num, hidden_dim
    CHECK_INPUT(cross_output_kernel, _st);      // layer_num, hidden_dim, hidden_dim
    CHECK_INPUT(cross_output_bias, _st);        // layer_num, hidden_dim
    CHECK_INPUT(ffn_layernorm_gamma, _st);      // layer_num, hidden_dim
    CHECK_INPUT(ffn_layernorm_beta, _st);       // layer_num, hidden_dim
    CHECK_INPUT(inter_kernel, _st);             // layer_num, hidden_dim, 4 * hidden_dim
    CHECK_INPUT(inter_bias, _st);               // layer_num, 4 * hidden_dim
    CHECK_INPUT(output_kernel, _st);            // layer_num, 4 * hidden_dim, hidden_dim
    CHECK_INPUT(output_bias, _st);              // layer_num, hidden_dim
    CHECK_INPUT(decoding_gamma, _st);           // hidden_dim
    CHECK_INPUT(decoding_beta, _st);            // hidden_dim
    CHECK_INPUT(embedding_table, _st);          // vocab_size, hidden_dim
    CHECK_INPUT(position_encoding_table, _st);  // max_step, hidden_dim
    CHECK_INPUT(embedding_kernel, _st);         // hidden_dim, vocab_size
    CHECK_INPUT(embedding_bias, _st);           // vocab_size

    CHECK_INPUT(pre_attr_layernorm_gamma, _st);          // hidden_dim
    CHECK_INPUT(pre_attr_layernorm_beta, _st);           // hidden_dim
    CHECK_INPUT(q_kernel, _st);                          // hidden_dim, hidden_dim
    CHECK_INPUT(q_bias, _st);                            // hidden_dim
    CHECK_INPUT(k_kernel, _st);                          // hidden_dim, hidden_dim
    CHECK_INPUT(k_bias, _st);                            // hidden_dim
    CHECK_INPUT(v_kernel, _st);                          // hidden_dim, hidden_dim
    CHECK_INPUT(v_bias, _st);                            // hidden_dim
    CHECK_INPUT(attr_output_kernel, _st);                // hidden_dim, hidden_dim
    CHECK_INPUT(attr_output_bias, _st);                  // hidden_dim
    CHECK_INPUT(pre_ffn_layernorm_gamma, _st);           // hidden_dim
    CHECK_INPUT(pre_ffn_layernorm_beta, _st);            // hidden_dim
    CHECK_INPUT(inter_kernel, _st);                      // 4 * hidden_dim, hidden_dim
    CHECK_INPUT(inter_bias, _st);                        // 4 * hidden_dim
    CHECK_INPUT(output_kernel, _st);                     // hidden_dim, 4 * hidden_dim
    CHECK_INPUT(output_bias, _st);                       // hidden_dim
    CHECK_INPUT(post_transformer_layernorm_gamma, _st);  // hidden_dim
    CHECK_INPUT(post_transformer_layernorm_beta, _st);   // hidden_dim
    std::cout<<"fastertransformerencoder init"<<std::endl;
    switch (_st) {
        case at::ScalarType::Float:
            std::cout<<"FP32 Mode"<<std::endl;
            ftencoder = new FTEncoder<float>(
                head_num, size_per_head, inter_size, enc_layer_num, allow_gemm_test, sparse, q_scaling, weights_enc);
            ftdecoding = new torch_ext::FTDecoding<float>(head_num,
                                                          size_per_head,
                                                          inter_size,
                                                          mem_hidden_dim,
                                                          layer_num,
                                                          vocab_size,
                                                          start_id,
                                                          end_id,
                                                          (float)beam_search_diversity_rate,
                                                          top_k,
                                                          (float)top_p,
                                                          (float)temperature,
                                                          (float)len_penalty,
                                                          (float)repetition_penalty,
                                                          weights_dec);

            break;
        case at::ScalarType::Half:
            std::cout<<"FP16 Mode"<<std::endl;
            ftencoder = new FTEncoder<half>(
                head_num, size_per_head, inter_size, enc_layer_num, allow_gemm_test, sparse, q_scaling, weights_enc);
            ftdecoding = new torch_ext::FTDecoding<half>(head_num,
                                                         size_per_head,
                                                         inter_size,
                                                         mem_hidden_dim,
                                                         layer_num,
                                                         vocab_size,
                                                         start_id,
                                                         end_id,
                                                         (float)beam_search_diversity_rate,
                                                         top_k,
                                                         (float)top_p,
                                                         (float)temperature,
                                                         (float)len_penalty,
                                                         (float)repetition_penalty,
                                                         weights_dec);
            break;
        default:
            throw std::runtime_error("Wrong Tensor type.");
    }   
    std::cout<< "receive enc_layer_num: "<<enc_layer_num<<std::endl;
    int_info_ = torch::empty({11}, torch::dtype(torch::kInt64));
    float_info_ = torch::empty({5}, torch::dtype(torch::kFloat64));
    int_info_[0] = head_num;
    int_info_[1] = size_per_head;
    int_info_[2] = inter_size;
    int_info_[3] = mem_hidden_dim;
    int_info_[4] = layer_num;
    int_info_[5] = vocab_size;
    int_info_[6] = start_id;
    int_info_[7] = end_id;
    int_info_[8] = top_k;

    float_info_[0] = beam_search_diversity_rate;
    float_info_[1] = top_p;
    float_info_[2] = temperature;
    float_info_[3] = len_penalty;
    float_info_[4] = repetition_penalty;

    head_info_ = torch::empty({7}, torch::dtype(torch::kInt64));
    head_info_[0] = head_num;
    head_info_[1] = size_per_head;
    head_info_[2] = (int64_t)remove_padding;
    head_info_[3] = enc_layer_num;
    head_info_[4] = (int64_t)allow_gemm_test;
    head_info_[5] = (int64_t)sparse;
    head_info_[6] = inter_size;
    scaling_info_ = torch::empty({1}, torch::dtype(torch::kFloat64));
    scaling_info_[0] = (double)q_scaling;

}

FasterTransformerNMT::~FasterTransformerNMT()
{
    delete ftencoder;
    delete ftdecoding;
}

th::Tensor FasterTransformerNMT::forward_enc(th::Tensor input,
                                                      th::Tensor sequence_lengths)
{
    CHECK_INPUT(input, _st);
    CHECK_TH_CUDA(sequence_lengths);
    CHECK_CONTIGUOUS(sequence_lengths);
    TORCH_CHECK(sequence_lengths.dtype() == torch::kInt32, "sequence_lengths dtype should be int32");
    size_t batch_size = (size_t)input.size(0);
    size_t seq_len = (size_t)input.size(1);

    auto output = torch::empty_like(input);
    ftencoder->forward(batch_size, seq_len, input, sequence_lengths, output, _remove_padding);
    return output;

    //th::Tensor output = fasterencoder->forward(input, sequence_lengths);
    //return output;
}
std::vector<th::Tensor> FasterTransformerNMT::forward_dec(int64_t beam_width,
                                                           int64_t max_seq_len,
                                                           th::Tensor memory,
                                                           th::Tensor memory_seq_lens)
{
    CHECK_INPUT(memory, _st);
    CHECK_TH_CUDA(memory_seq_lens);
    CHECK_CONTIGUOUS(memory_seq_lens);
    TORCH_CHECK(memory_seq_lens.dtype() == torch::kInt32, "mem_seq_lens dtype should be int32");

    int batch_size = (int)(memory.size(0) / beam_width);
    auto output_ids = torch::empty({batch_size * beam_width * max_seq_len},
                                   torch::dtype(torch::kInt32).device(torch::kCUDA).requires_grad(false));
    auto parent_ids = torch::empty({batch_size * beam_width * max_seq_len},
                                   torch::dtype(torch::kInt32).device(torch::kCUDA).requires_grad(false));
    auto out_seq_lens =
        torch::empty({batch_size * beam_width}, torch::dtype(torch::kInt32).device(torch::kCUDA).requires_grad(false));
    auto out_finals =
        torch::empty({batch_size * beam_width * 2}, torch::dtype(torch::kInt32).device(torch::kCUDA).requires_grad(false));

    auto out_scores =
        torch::empty({batch_size * beam_width}, torch::dtype(torch::kFloat32).device(torch::kCUDA).requires_grad(false));
    
    ftdecoding->forward(
        (size_t)beam_width, (size_t)max_seq_len, memory, memory_seq_lens, output_ids, parent_ids, out_seq_lens, out_finals,
        out_scores);
    return std::vector<th::Tensor>{output_ids, parent_ids, out_seq_lens, out_finals, out_scores};

    //std::vector<th::Tensor> decoding_output = fasterdecoding->forward(beam_width, max_seq_len, memory, memory_seq_lens);
    //return decoding_output;
}

std::vector<th::Tensor> FasterTransformerNMT::get_pickle_info() const
{
    std::vector<th::Tensor> tmp(weights_dec);
    std::vector<th::Tensor> tmp_enc(weights_enc);
    for(auto item : tmp_enc){
        tmp.push_back(item);
    }   
    tmp.push_back(int_info_);
    tmp.push_back(float_info_);
    tmp.push_back(head_info_);
    tmp.push_back(scaling_info_);
    return tmp;
}

}  // namespace torch_ext

static auto fasterTransformerNMTTHS =
#ifdef LEGACY_THS
    torch::jit::class_<torch_ext::FasterTransformerNMT>("FasterTransformerNMT")
#else
    torch::jit::class_<torch_ext::FasterTransformerNMT>("FasterTransformer", "NMT")
#endif
        .def(torch::jit::init<int64_t,
                              int64_t,
                              int64_t,
                              int64_t,
                              int64_t,
                              int64_t,
                              int64_t,
                              int64_t,
                              double,
                              int64_t,
                              double,
                              double,
                              double,
                              double,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                              th::Tensor,
                                int64_t,
                              bool,
                              bool,
                              bool,
                              double>())
        .def("forward_enc", &torch_ext::FasterTransformerNMT::forward_enc)
        .def("forward_dec", &torch_ext::FasterTransformerNMT::forward_dec)
        .def_pickle(
            [](const c10::intrusive_ptr<torch_ext::FasterTransformerNMT>& self) -> std::vector<th::Tensor> {
                return self->get_pickle_info();
            },
            [](std::vector<th::Tensor> state) -> c10::intrusive_ptr<torch_ext::FasterTransformerNMT> {
                int head_num = state[46][0].item().to<int>();
                int size_per_head = state[46][1].item().to<int>();
                int inter_size = state[46][2].item().to<int>();
                int mem_hidden_dim = state[46][3].item().to<int>();
                int layer_num = state[46][4].item().to<int>();
                int vocab_size = state[46][5].item().to<int>();
                int start_id = state[46][6].item().to<int>();
                int end_id = state[46][7].item().to<int>();
                int top_k = state[46][8].item().to<int>();

                // TODO(bhsueh) Here may have bugs
                double beam_search_diversity_rate = state[51][0].item().to<double>();
                double top_p = state[51][1].item().to<double>();
                double temperature = state[51][2].item().to<double>();
                double len_penalty = state[51][3].item().to<double>();
                double repetition_penalty = state[51][4].item().to<double>();

                bool remove_padding = false;
                bool allow_gemm_test = false;
                bool sparse = false;
                double q_scaling = 1.0;
                int64_t enc_layer_num = 20;

                return c10::make_intrusive<torch_ext::FasterTransformerNMT>(head_num,
                                                                                 size_per_head,
                                                                                 inter_size,
                                                                                 mem_hidden_dim,
                                                                                 layer_num,
                                                                                 vocab_size,
                                                                                 start_id,
                                                                                 end_id,
                                                                                 beam_search_diversity_rate,
                                                                                 top_k,
                                                                                 top_p,
                                                                                 temperature,
                                                                                 len_penalty,
                                                                                 repetition_penalty,
                                                                                 state[0],
                                                                                 state[1],
                                                                                 state[2],
                                                                                 state[3],
                                                                                 state[4],
                                                                                 state[5],
                                                                                 state[6],
                                                                                 state[7],
                                                                                 state[8],
                                                                                 state[9],
                                                                                 state[10],
                                                                                 state[11],
                                                                                 state[12],
                                                                                 state[13],
                                                                                 state[14],
                                                                                 state[15],
                                                                                 state[16],
                                                                                 state[17],
                                                                                 state[18],
                                                                                 state[19],
                                                                                 state[20],
                                                                                 state[21],
                                                                                 state[22],
                                                                                 state[23],
                                                                                 state[24],
                                                                                 state[25],
                                                                                 state[26],
                                                                                 state[27],
                                                                                 state[28],
                                                                                 state[29],
                                                                                 state[30],
                                                                                 state[31],
                                                                                 state[32],
                                                                                 state[33],
                                                                                 state[34],
                                                                                 state[35],
                                                                                 state[36],
                                                                                 state[37],
                                                                                 state[38],
                                                                                 state[39],
                                                                                 state[40],
                                                                                 state[41],
                                                                                 state[42],
                                                                                 state[43],
                                                                                 state[44],
                                                                                 state[45],
                                                                                 enc_layer_num,
                                                                                 remove_padding,
                                                                                 allow_gemm_test,
                                                                                 sparse,
                                                                                 q_scaling);
            });
```

**Comment by byshiue** [comment]
You can try printing some intermediate buffer in FT, comparing under FP32 and FP16.

**Comment by SefaZeng** [comment]
> You can try printing some intermediate buffer in FT, comparing under FP32 and FP16.

Hi, @byshiue Hope you are still here. This may be the last question. I have share the parameter of encoder embedding and decoder embedding and it works well. I am trying to share `decoder embedding` and `output projection` in decoding but the results are wrong.
I notice that T5 shares the parameter for decoder embedding and output projection by setting the `decoding_weights.pre_decoder_embedding_table`  and `decoding_weights.post_decoder_embedding.kernel` to the same weight. But as I know, the shapes of these two matrices are transposed to each other.  `pre_decoder_embedding_table` should be a `[vocab_size, embed_dim]` matrix and `post_decoder_embedding.kernel` should be a `[embed_dim, vocab_size]` matrix. So it should transpose the `post_decoder_embedding.kernel` when it comes to output. But I can't find how T5 does that. 
Any advice is appreciated. Thx~

**Comment by SefaZeng** [comment]
> You can try printing some intermediate buffer in FT, comparing under FP32 and FP16.

Hi, @byshiue. I can now share these 3 embeddings and get reasonable results, but there are two problems that I found.
1. `vocab_padding` leads to wrong results. Since my vocab size is 251095, when using fp16, `vocab_padding` causes the `vocab size` to change to 251096, and then I find that the weights change after padding. Specifically, the wrong weight appears after the pointer goes through a word table size, but I'm not sure what it is causing.
2. Because of what I mentioned above, a transpose operation is needed to do the computation after embedding sharing. So I replaced CUBLAS_OP_N with CUBLAS_OP_T, which made the time consumption increase significantly.The original translation time of a sentence with 98 tokens is 0.78s, but after using transposition, it becomes 3.3s, which is unacceptable. I am not sure if there is something wrong with my operation.

Any advice would be appreciated. Thx in advancing.
